iavl/iavl_range_proofs.go

```diff
@@ -0,0 +1,419 @@
+			return
+		}
+	} else if proof.Left != nil && proof.Left.Path.isRightmost() {
+		// No key found. Range starts outside of the right boundary.
+		return
+	} else if proof.Right != nil && proof.Right.Path.isLeftmost() {
+		// No key found. Range ends outside of the left boundary.
+		return
+	} else if proof.Left != nil && proof.Right != nil &&
+		proof.Left.Path.isLeftAdjacentTo(proof.Right.Path) {
+		// No key found. Range is between two existing keys.
+		return
+	}
+
+	return ErrInvalidProof
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+// KeyLastInRangeProof is a proof that a given key is the last in a given range.
+type KeyLastInRangeProof struct {
+	KeyExistsProof `json:"key_proof"`
+
+	Left  *PathWithNode `json:"left"`
+	Right *PathWithNode `json:"right"`
+}
+
+// String returns a string representation of the proof.
+func (proof *KeyLastInRangeProof) String() string {
+	// TODO(cloudhead): Needs work.
+	return fmt.Sprintf("%#v", proof)
+}
+
+// Verify that the last in range proof is valid.
+func (proof *KeyLastInRangeProof) Verify(startKey, endKey, key, value []byte, root []byte) (err error) {
+	if key != nil && (bytes.Compare(key, startKey) == -1 || bytes.Compare(key, endKey) == 1) {
+		return ErrInvalidInputs
+	}
+	if proof.Left == nil && proof.Right == nil && proof.PathToKey == nil {
+		return ErrInvalidProof
+	}
+	if err := verifyPaths(proof.Left, proof.Right, startKey, endKey, root); err != nil {
+		return err
+	}
+
+	if proof.PathToKey != nil {
+		if err := proof.KeyExistsProof.Verify(key, value, root); err != nil {
+			return err
+		}
+		if bytes.Equal(key, endKey) {
+			return
+		}
+		if proof.PathToKey.isRightmost() {
+			return
+		}
+		if proof.Right != nil &&
+			proof.PathToKey.isLeftAdjacentTo(proof.Right.Path) {
+			return
+		}
+	} else if proof.Right != nil && proof.Right.Path.isLeftmost() {
+		return
+	} else if proof.Left != nil && proof.Left.Path.isRightmost() {
+		return
+	} else if proof.Left != nil && proof.Right != nil &&
+		proof.Left.Path.isLeftAdjacentTo(proof.Right.Path) {
+		return
+	}
+	return ErrInvalidProof
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+// KeyRangeProof is proof that a range of keys does or does not exist.
+type KeyRangeProof struct {
+	RootHash   data.Bytes   `json:"root_hash"`
+	PathToKeys []*PathToKey `json:"paths"`
+
+	Left  *PathWithNode `json:"left"`
+	Right *PathWithNode `json:"right"`
+}
+
+// Verify that a range proof is valid.
+//
+// This method expects the same parameters passed to query the range.
+func (proof *KeyRangeProof) Verify(
+	startKey, endKey []byte, limit int, keys, values [][]byte, root []byte,
+) error {
+	if len(proof.PathToKeys) != len(keys) || len(values) != len(keys) {
+		return ErrInvalidInputs
+	}
+	if limit > 0 && len(keys) > limit {
+		return ErrInvalidInputs
+	}
+	if len(proof.PathToKeys) == 0 && proof.Left == nil && proof.Right == nil {
```

Comment:
This condition makes me wonder whether we shouldn't just terminate right at the top with check `if len(proof.PathToKeys) == 0` since the first condition that we are dealing with is `if len(proof.PathToKeys) != len(keys)` of which we can only get here if `len(keys) == 0`. What do we do if `len(keys) == 0`?